x-pack/plugins/cases/server/connectors/cases/cases_connector_executor.test.ts

```diff
@@ -0,0 +1,2085 @@
+
+      it('gets the oracle records correctly', async () => {
+        await connectorExecutor.execute({
+          alerts,
+          groupingBy: [],
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        });
+
+        expect(mockBulkGetRecords).toHaveBeenCalledWith(['so-oracle-record-0']);
+      });
+    });
+
+    describe('Cases', () => {
+      it('generates the case ids correctly', async () => {
+        await connectorExecutor.execute({
+          alerts,
+          groupingBy: [],
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        });
+
+        expect(mockGetCaseId).toHaveBeenCalledTimes(1);
+
+        expect(mockGetCaseId).nthCalledWith(1, {
+          ruleId: rule.id,
+          grouping: {},
+          owner,
+          spaceId: 'default',
+          counter: 1,
+        });
+      });
+
+      it('gets the cases correctly', async () => {
+        await connectorExecutor.execute({
+          alerts,
+          groupingBy: [],
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        });
+
+        expect(casesClientMock.cases.bulkGet).toHaveBeenCalledWith({
+          ids: ['mock-id-1'],
+        });
+      });
+    });
+
+    describe('Alerts', () => {
+      it('attach all alerts to the same case when the grouping is not defined', async () => {
+        await connectorExecutor.execute({
+          alerts,
+          groupingBy: [],
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        });
+
+        expect(casesClientMock.attachments.bulkCreate).toHaveBeenCalledTimes(1);
+
+        expect(casesClientMock.attachments.bulkCreate).nthCalledWith(1, {
+          caseId: 'mock-id-1',
+          attachments: alerts.map((alert) => ({
+            alertId: alert._id,
+            index: alert._index,
+            owner: 'securitySolution',
+            rule: {
+              id: 'rule-test-id',
+              name: 'Test rule',
+            },
+            type: 'alert',
+          })),
+        });
+      });
+    });
+  });
+
+  describe('Retries', () => {
+    it('attaches the alerts correctly when bulkGetRecords fails', async () => {
+      mockBulkGetRecords
+        .mockResolvedValueOnce([
+          {
+            id: groupedAlertsWithOracleKey[2].oracleKey,
+            type: CASE_ORACLE_SAVED_OBJECT,
+            message: 'getting records: mockBulkGetRecords error',
+            statusCode: 409,
+            error: 'Conflict',
+          },
+        ])
+        .mockResolvedValueOnce(oracleRecords);
+
+      await expect(() =>
+        connectorExecutor.execute({
+          alerts,
+          groupingBy,
+          owner,
+          rule,
+          timeWindow,
+          reopenClosedCases,
+        })
+      ).rejects.toThrowErrorMatchingInlineSnapshot(
+        `"Conflict: getting records: mockBulkGetRecords error"`
+      );
+
+      resetCounters();
```

Comment:
The generation of the IDs is mocked. It uses counters to get an incremental ID each time an ID is requested. We need to reset the counters before we retry the execution.